sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -56,29 +56,49 @@ case class CreateArray(children: Seq[Expression]) extends Expression {
   }
 
   override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val unsafeArrayClass = classOf[UnsafeArrayData].getName
     val arrayClass = classOf[GenericArrayData].getName
     val values = ctx.freshName("values")
     ctx.addMutableState("Object[]", values, s"this.$values = null;")
 
-    ev.copy(code = s"""
-      final boolean ${ev.isNull} = false;
-      this.$values = new Object[${children.size}];""" +
-      ctx.splitExpressions(
-        ctx.INPUT_ROW,
-        children.zipWithIndex.map { case (e, i) =>
-          val eval = e.genCode(ctx)
-          eval.code + s"""
+    val ArrayType(dt, _) = dataType
+    val evals = children.map(e => e.genCode(ctx))
+    val isPrimitiveArray = ctx.isPrimitiveType(dt) && evals.forall(_.isNull == "false")
```

Comment:
Is it possible that a child expression is nullable but its evaluation's `isNull == "false"`?

I think it is. Because `isNull` of a nullable expression could be `true` or `false`.